staging/src/k8s.io/kms/encryption/managed.go

```diff
@@ -0,0 +1,264 @@
+// a cache, a reference to an remote cipher (like a KMS service or HMS) and
+// does an initial encryption call to the remote cipher.
+func NewManagedCipher(remoteCipher EncrypterDecrypter) (*ManagedCipher, error) {
+	if remoteCipher == nil {
+		klog.Infof("create managed cipher without remote encryption failed")
+		return nil, ErrNoCipher
+	}
+
+	cipher, err := NewAESGCM()
+	if err != nil {
+		klog.Infof("create new cipher: %w", err)
+		return nil, err
+	}
+
+	keyID, encCipher, err := remoteCipher.Encrypt(cipher.Key())
+	if err != nil {
+		klog.Infof("encrypt with remote: %w", err)
+		return nil, err
+	}
+
+	cache := newCache(cacheSize)
+	cache.Add(encCipher, cipher)
+
+	mc := ManagedCipher{
+		keys:               cache,
+		counter:            0,
+		expires:            time.Now().Add(week),
+		remoteKMS:          remoteCipher,
+		currentRemoteKMSID: keyID,
+		currentLocalKEK:    encCipher,
+	}
+
+	klog.Infof("new managed cipher is created")
+
+	go func() {
+		if err := mc.addFallbackCipher(); err != nil {
+			klog.Infof("set up fallback cipher: %w", err)
+		}
+	}()
+
+	return &mc, nil
+}
+
+func (m *ManagedCipher) addFallbackCipher() error {
+	cipher, err := NewAESGCM()
+	if err != nil {
+		klog.Infof("create new currentCipher: %w", err)
+		return err
+	}
+
+	keyID, encCipher, err := m.remoteKMS.Encrypt(cipher.Key())
+	if err != nil {
+		klog.Infof("encrypt with remote: %w", err)
+		return err
+	}
+
+	m.keys.Add(encCipher, cipher)
+	m.fallbackLocalKEK = encCipher
+	m.fallbackRemoteKMSID = keyID
+
+	return nil
+}
+
+func (m *ManagedCipher) manageKey() error {
+	m.m.Lock()
+	defer m.m.Unlock() // should be bound to main thread
+
+	// If the key is safe to use, do nothing.
+	m.counter = m.counter + 1
+	if m.counter < MaxUsage && time.Now().Before(m.expires) {
+		return nil
+	}
+
+	// If fallback cipher doesn't exist, create one synchronously.
+	if m.fallbackLocalKEK == nil {
+		// In case that an error happened, while setting the fallback cipher
+		// asynchronously, do it now synchronously.
+		if err := m.addFallbackCipher(); err != nil {
+			return err
+		}
+	}
+
+	// Switch from current to fallback cipher
+	m.currentRemoteKMSID = m.fallbackRemoteKMSID
+	m.fallbackRemoteKMSID = nil
+	m.currentLocalKEK = m.fallbackLocalKEK
+	m.fallbackLocalKEK = nil
+
+	// Reset checks.
+	m.expires = time.Now().Add(week)
+	m.counter = 0
+
+	go func() {
+		// Add fallback cipher asynchronously and optimistically.
+		if err := m.addFallbackCipher(); err != nil {
+			klog.Infof("set up fallback cipher: %w", err)
+		}
+	}()
+
+	return nil
+}
+
+// Encrypt encrypts given plaintext and returns the key used in encrypted form.
+// The encrypted key is encrypted by the given remote KMS.
+func (m *ManagedCipher) Encrypt(pt []byte) ([]byte, []byte, []byte, error) {
+	if err := m.manageKey(); err != nil {
+		return nil, nil, nil, fmt.Errorf("manage keys upfront of an encryption: %w", err)
+	}
+
+	cipher, ok := m.keys.Get(m.currentLocalKEK)
```

Comment:
Based on the current implementation, you cannot access this field without holding a lock.